### PR TITLE
Update UITableView to track changes precisely

### PR DESCRIPTION
**Pull request by swankjesse**
Previously we called reloadData() whenever anything changed in the underlying model.

With this change we fire precise events to the UI when inserts and removes happen in the source LazyList. This attempts to coalesce changes to itemsBefore() and itemsAfter() with adjacent changes and the beginning and end of the items list. To implement this the changes are buffered into an intermediate model before they are applied.

When a cell changes from being a placeholder to a loaded item, or the opposite, this does a cell content change without firing an event through the UITableView. To implement this we must track which cells are currently displaying placeholders. This PR includes a new datastructure, SparseList, to implement this tracking.

This PR also changes the number of placeholders that guest code offers to host code. I found experimentally that 20 placeholders was not enough for UITableView.

This PR changes LazyList from UICollectionView to
UITableView. This change was potentially unnecessary, though UITableView has fewer features that we don't need.

### CODE REVIEW

File: redwood-lazylayout-widget/src/commonMain/kotlin/app/cash/redwood/lazylayout/widget/LazyListUpdateProcessor.kt
Comment (swankjesse) [inline redwood-lazylayout-widget/src/commonMain/kotlin/app/cash/redwood/lazylayout/widget/LazyListUpdateProcessor.kt:90]: Need to test fencepost logic on these.

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (veyndan) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:67]: The goal of `WindowedLazyList` was to have `UIViewLazyList` and `ViewLazyList` to share the bulk of their implementations. They are fairly similar (e.g., view based, platform provides a recycling list view that allows atomic updates). The integration of `WindowedLazyList` with `UIViewLazyList` wasn't complete, namely that `UICollectionViewListUpdateCallback` didn't yet delegate to atomic updates.

What are your thoughts on continuing down that path vs switching approach as you did here?

**Review by veyndan — CHANGES_REQUESTED** [review]
Just a drive-by review for now:

> 

> This PR changes LazyList from UICollectionView to UITableView. This change was potentially unnecessary, though UITableView has fewer features that we don't need.

Switching it back to `UITableView` means that `LazyRow` can no longer work (see https://github.com/cashapp/redwood/pull/942#discussion_r1172570297). I'd be very hesitant to break support for this!

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (veyndan) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:67]: There's this issue from last week that has more info: https://github.com/cashapp/redwood/issues/1498

**Comment by veyndan** [comment]
> This PR also changes the number of placeholders that guest code offers to host code. I found experimentally that 20 placeholders was not enough for UITableView.

The underlying issue is probably https://github.com/cashapp/redwood/issues/1500. Increasing to 30 placeholders increases the number of serialized views, which'll be unnecessary for the majority of lists.

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (dnagler) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:83]: Calling `indexPathForItem` this frequently might be expensive--it might be worth trying to just create our own index paths, and enforcing that the table view only has a single section (so that every index path is initialized with `.init(section: 0, row: index)`

Comment (dnagler) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:83]: nit: is it possible to use the more modern/bridged `IndexPath` type here?

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (dnagler) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt]: nit: You might want to add an assertion here that `section == 0`

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (dnagler) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt]: nit: is this needed? 44 is the default value since iOS 11

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (dnagler) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:167]: Do we plan to support this in the future, or can/should we punt this indefinitely? (Has this already been built on the Android side?)

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (swankjesse) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:67]: This new class started out as a copy-paste of WindowedLazyList, addressing problems on iOS by adding precise updates.

Are precise updates useful for RecyclerView? If precise updates will make RecyclerView faster, then we should follow-up by migrating RecyclerView to LazyListUpdateProcessor (and delete the old WindowedLazyList). If precise updates are not necessary for RecyclerView, then we should not do precise updates; it’s a bit of work to provide them.

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (dnagler) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:268]: nit: since `widgetView` is a subview of `contentView`, I would probably do
```
contentView.setFrame(bounds)
widgetView.setFrame(contentView.bounds)
```
instead, but it should have the same effect

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (dnagler) [inline redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:271]: this can probably just return `contentView.sizeThatFits(size)`

**Comment by swankjesse** [comment]
@veyndan ooooh I like that technique to handle lack of placeholders. Right now the app crashes if there aren’t enough placeholders. I’d like to introduce this performance regression to fix that crash, and then later we can implement #1500 as an optimization.

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (swankjesse) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:67]: Reading that issue makes it clear to me that this new thing will benefit both platforms. I’d still like to land it one-platform-at-a-time, mostly because our test coverage here is lacking.

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (swankjesse) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:167]: I’d like to see what happens if I swap out UITableView for UICollectionView. If I’m lucky it’ll just be tracking slight differences in the APIs with no performance impact. I think I’ll be lucky.

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (swankjesse) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:83]: I couldn’t find it in the Kotlin/Native APIs.

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (swankjesse) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt]: Done.

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (dnagler) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:268]: we'll also definitely want to call `widgetView.sizeToFit()` here

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (swankjesse) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt]: Fixed!

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (swankjesse) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:268]: After playing with this I’ve learned that contentView is disinterested in delegating to its children for stuff like sizing, so the cell needs to control both. I’m gonna keep it as-is for now ’cause it seems to be working!

**Comment by dnagler** [comment]
After pairing on this (and seeing marked improvements over current performance!!), I'm in favor of landing this as-is and continuing to iterate, especially since this will be a great help to continuing our QA process on the iOS side (right now our QA team is testing a very broken experience).

**Review by dnagler — APPROVED** [review]

**Comment by swankjesse** [comment]
@veyndan

I spent a couple days trying to switch this back to UICollectionView and I’ve been unsuccessful thus far. I can’t figure out how to get dynamic heights working without it exploding because it runs out of placeholders.

Could you please release your merge-block on this PR? We can do the UITableView to UICollectionView switch in a follow-up, but I’d like to unblock the stuff we need right now in Cash App.

File: redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyList.kt
Comment (veyndan) [inline redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyList.kt]: Revert? This just change just affects Android, as iOS doesn't know how to request more placeholders.

File: redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyList.kt
Comment (veyndan) [inline redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyList.kt:54]: Can we add a `TODO` to drop this down to 20 again? It's really just blocked on iOS being unable to request more placeholders.

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (veyndan) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:67]: That sounds good. Do you mind adding a `TODO` here (or manually create an issue) so that we remember that these two mechanisms should ideally be merged?

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (veyndan) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:167]: @dnagler To answer your question directly:

1. We plan to support this in the future. We'll need this within Cash for things like the QAB, but also, `LazyRow` should be provided for external consumers (i.e., non Cash people), as these components are supposed to be the lazy counterparts to the existing `Column`/`Row` components.
2. Yes, it's already been built on the Android side.

**Review by veyndan — APPROVED** [review]
Stamping to unblock

File: redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyList.kt
Comment (swankjesse) [reply redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyList.kt]: Done.

File: redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyList.kt
Comment (swankjesse) [reply redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyList.kt:54]: Done.

```
  // TODO(jwilson): drop this down to 20 once this is fixed:
  //     https://github.com/cashapp/redwood/issues/1551
```

File: redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt
Comment (swankjesse) [reply redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt:67]: Here’s a tracking issue: https://github.com/cashapp/redwood/issues/1552

**Pull request merged**
